### PR TITLE
[WIP] Fix #160 - 64-bit-execution-via-heavens-gate

### LINKED ISSUE #160: syscall-using-heavens-gate
## Prerequisites

* [X] Put an X between the brackets on this line if you have done all of the following:
    * Checked that your rule idea isn't already filed: [search](https://github.com/fireeye/capa-rules/issues?q=is%3Aissue+is%3Aopen+)

## Summary

Would it make sense to write a rule that looks for a function that syscall's and leverages the heaven's gate technique?

## Examples

From: https://download.bitdefender.com/resources/files/News/CaseStudies/study/318/Bitdefender-TRR-Whitepaper-Maze-creat4351-en-EN-GenericUse.pdf

image

Sample: https://www.virustotal.com/gui/file/953d574ef0f49e886e270c2887586aab5c0d37616dd99f422838a1257abdbe3b/community

## Features

Maybe the following with function scope:
```
push 0x33
retf
syscall
```
and/or
```
syscall
move 0x23
retf
```

## Additional context

N/A

### Namespace

Not sure...

### References

N/A

### Other rule meta information

N/A

### PR DESCRIPTION
**Pull request by recvfrom**
Proposed rule for #160 

See https://github.com/fireeye/capa-testfiles/pull/35 for test file

### CODE REVIEW

File: executable/pe/64-bit-execution-via-heavens-gate.yml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+      - and:
+        - or:
+          - mnemonic: push = 'push 33h'
+          - mnemonic: mov = cover any mov ESP / EBP equivalents
+        - number: 0x33 = set up retf to push 0x33 to CS indicating 64-bit mode
+      - mnemonic: call = 'call $+5' pushes the current EIP onto the stack, +5 to jump past call insn bytes
+      - and:
+        - mnemonic: add = 'add dword ptr[esp], 5' updates the return address to point after retf
+        - number: 0x5 = length of add + retf insn bytes
+      - mnemonic: retf = set EIP = [ESP] and CS = [ESP+4]
```
Comment (williballenthin) [inline executable/pe/64-bit-execution-via-heavens-gate.yml:24]: can we add a BB count limit to the function, too, to help tighten it up?

File: executable/pe/64-bit-execution-via-heavens-gate.yml [hunk 2]
```diff
@@ -0,0 +1,24 @@
+rule:
+  meta:
+    name: 64-bit-execution-via-heavens-gate
+    namespace: executable/pe
```
Comment (williballenthin) [inline executable/pe/64-bit-execution-via-heavens-gate.yml:4]: lets put this under anti-analysis somewhere, maybe anti-disasm?

File: executable/pe/64-bit-execution-via-heavens-gate.yml [hunk 3]
```diff
@@ -0,0 +1,24 @@
+rule:
+  meta:
+    name: 64-bit-execution-via-heavens-gate
```
Comment (williballenthin) [inline executable/pe/64-bit-execution-via-heavens-gate.yml:3]: filename uses dashes, but the rule name should use spaces and be ready for human display

**Review by williballenthin — CHANGES_REQUESTED** [review]
couple comments, nothing major, definitely excited for this!

File: executable/pe/64-bit-execution-via-heavens-gate.yml (hunk 1 above)
Comment (recvfrom) [reply executable/pe/64-bit-execution-via-heavens-gate.yml:24]: good thinking!

**Review by williballenthin — APPROVED** [review]
lgtm

**Comment by williballenthin** [comment]
thank you!

**Comment by recvfrom** [comment]
awesome, thanks!

@williballenthin would you mind adding a label called `hacktoberfest-accepted` to this?  I'm participating in Hacktoberfest 2020 and it'd be cool if this PR could count for it. Thanks!

https://hacktoberfest.digitalocean.com/details#details

**Comment by williballenthin** [comment]
of course, done!

hope we'll still see you in November and beyond, though ;-)

**Pull request merged**